Stricter checks for SDF sniffing

nsoranzo opened the pull request:
I had a bunch of protein FASTA files being sniffed as 'sdf'.

Also fix https://github.com/galaxyproject/galaxy/issues/4558 .

nsoranzo committed:
Stricter checks for SDF sniffing

I had a bunch of protein FASTA files being sniffed as 'sdf'.

Also fix https://github.com/galaxyproject/galaxy/issues/4558 .

bgruening commented on lib/galaxy/datatypes/molecules.py:
```diff
@@ -101,17 +105,27 @@ def sniff(self, filename):
-        else:
-            return False
+        m_end_found = False
+        limit = 500
```
This probably needs to be increased. I have seen molecules with 1000 atoms, so maybe 1500?

nsoranzo replied:
Good point! I've increased this to 10000, but also added a strict check on the 4th line which would stop the method early for most files.

bgruening reviewed (commented):
Thanks @nsoranzo! LGTM, small comment included.

nsoranzo committed:
More precise SDF sniffer

First check 4th line: if this is conforming, scan up to 10000 lines.
Add 2 test files downloaded from:

https://github.com/rdkit/rdkit/tree/master/Code/GraphMol/FileParsers/test_data

jmchilton commented on lib/galaxy/datatypes/molecules.py:
```diff
@@ -92,26 +92,53 @@ def sniff(self, filename):
+        limit = 10000
+        idx = 0
+        with open(filename) as in_file:
+            for line in in_file:
```
We really need to stop doing this - this line can consume unbounded memory on relatively sane user supplied inputs. I guess there are a lot of instances of it though and I'm not finding my previous advice on fixing it. (I believe the right thing to do is pick a maximum line size and use readlines directly instead of the iterator).

nsoranzo replied:
xref. https://github.com/galaxyproject/galaxy/issues/4319

jvolkening commented:
Will this make it into 18.01? I'm having FASTQ files with '$$$$' in qualities detected as SDF.

nsoranzo commented:
@jvolkening Yes, it will.

mvdbeek commented:
I guess we should backport this change if the problem appears on 17.09.
